docs/migration-recipe.md

```diff
@@ -0,0 +1,117 @@
+----------------------
+
+### Prepare salt configuration for migration:  
+
+1.  Ensure that salt-master, loadbalancer, and host in question are up, and that their health check for the relevant service is failing in haproxy after the host is fully up
+
+To view haproxy status: 
+
+-vagrant up the salt-master, loadbalancer, and host in question (`vagrant up`)
+
+-Prepare an ssh configuration file to access the host with native ssh commands: `vagrant ssh-config salt-master loadbalancer >> vagrant-ssh` 
+
+-Open an ssh session with port forwarding to the haproxy status page: `ssh -L 4646:127:0.0.1:464 -F vagrant-ssh loadbalancer`
+
+-view the haproxy status page in your browser `loacalhost:4646/haproxy?stats` 
+
+2.  Edit pillar data for roles.sls to include both old and new hostnames (ex. hostname*)
+
+### Migrate the host:
+
+#### Update Salt Master with latest config including prep from above
+
+1.  ssh into the salt-master server `ssh salt-master`
+
+2.  Navigate to srv/psf-salt `user@salt:~$ cd srv/psf-salt`
+
+3.  Run `user@salt:/srv/psf-salt$ sudo git pull`
+
+4.  Run highstate to update the roles settings to reflect the new matchng pattern, as well as additional changes to support migration: `user@salt:/srv/psf-salt$ sudo salt-call state.highstate`
+
+#### Ensure new configuration doesn't impact host being migrated
+
+1.  ssh into the old host `ssh hostname`
+
+2.  Run `user@old-host:~$ sudo salt-call state.highstate`
+
+#### Create new host:
+
+1.  Start a new droplet  in digital ocean, and check resources being used on old host to see if we are over or under spending on resources 
+
+2.  Create a new droplet with new version of ubuntu, appropriate resources,  and name it according to hostname + 2004
+
+#### Provision new host for migration:
+
+1. ssh into new host `ssh new-host`
+
+2.  Add Salt repositories for our current target version (add the apt-repo and install salt-minion package)
+
+-   `user@new-host:~$ wget --quiet -O /usr/share/keyrings/salt-archive-keyring.gpg https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004/salt-archive-keyring.gpg`
+
+-  `user@new-host:~$ echo "deb [signed-by=/usr/share/keyrings/salt-archive-keyring.gpg arch=$(dpkg --print-architecture)] https://repo.saltproject.io/py3/ubuntu/20.04/$(dpkg --print-architecture)/3004 focal main" > /etc/apt/sources.list.d/salt.list`
+
+3.  Install and configure the salt-minion. On new host, run the command
+
+- `user@new-host:~$ apt-get update -y && apt-get install -y --no-install-recommends salt-minion`
+
+- On the original host, look through `etc/salt/minion.d*` to set up salt-minion configuration files to match on new host 
+
+4. Restart the salt-minion service on the new host to pickup the configuration and register with salt-master: `user@new-host:~$ sudo salt-call service.restart`
+
+5.  On salt-master, accept the key for the new host: `user@new-host:~$ sudo salt-key -a hostname`
+
+6.  Run `user@new-host:~$ sudo salt-call state.highstate`
```

Comment:
An important check at this point is to ensure that the new host _is not_ serving traffic since no data has been migrated.

I suggest a 7th step of:

"7. Ensure that the new host is not passing health checks in the loadbalancer: `ssh -L 4646:127.0.0.1:4646 lb-a.nyc1.psf.io` then open http://localhost:4646/haproxy?stats in your browser.

One other thing to "finalize" a new host, is to make sure that the salt-master has run highstate after the the new-host has finished a high state for the first time.

I suggest a 8th step, of :

"8. Run highstate on the salt-master to create public dns record for the new-host `user@salt:/srv/psf-salt$ sudo salt-call state.highstate`"